src/Microsoft.Health.Dicom.Functions/Migration/MigrationFilesDurableFunction.Activity.cs

```diff
@@ -0,0 +1,110 @@
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>
+    /// A task representing the asynchronous get operation. The value of its <see cref="Task{TResult}.Result"/>
+    /// property contains a list of batches as defined by their smallest and largest watermark.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="arguments"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(GetInstanceBatchesByTimeStampAsync))]
+    public Task<IReadOnlyList<WatermarkRange>> GetInstanceBatchesByTimeStampAsync(
+        [ActivityTrigger] MigrationBatchCreationArguments arguments,
+        ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        if (arguments.MaxWatermark.HasValue)
+        {
+            logger.LogInformation("Dividing up the instances into batches starting from the largest watermark {Watermark}.", arguments.MaxWatermark);
+        }
+        else
+        {
+            logger.LogInformation("Dividing up the instances into batches starting from the end.");
+        }
+
+        return _instanceStore.GetInstanceBatchesByTimeStampAsync(
+            arguments.BatchSize,
+            arguments.MaxParallelBatches,
+            IndexStatus.Created,
+            arguments.StartFilterTimeStamp,
+            arguments.EndFilterTimeStamp,
+            arguments.MaxWatermark,
+            CancellationToken.None);
+    }
+
+    /// <summary>
+    /// Asynchronously migrate frame range files
+    /// </summary>
+    /// <param name="arguments">The options that include the instances to re-index and the query tags.</param>
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>A task representing the <see cref="MigrateFrameRangeFilesAsync"/> operation.</returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="arguments"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(MigrateFrameRangeFilesAsync))]
+    public async Task MigrateFrameRangeFilesAsync([ActivityTrigger] MigrationFilesBatchArguments arguments, ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        IReadOnlyList<VersionedInstanceIdentifier> instanceIdentifiers =
+            await _instanceStore.GetInstanceIdentifiersByWatermarkRangeAsync(arguments.WatermarkRange, IndexStatus.Created);
+
+        var versions = instanceIdentifiers.Select(x => x.Version).ToList();
+
+        await Parallel.ForEachAsync(
+            versions,
+            new ParallelOptions
+            {
+                CancellationToken = default,
+                MaxDegreeOfParallelism = _options.MaxParallelThreads,
+            },
+            async (version, token) =>
```

Comment:
nit: these functions don't need `async` and `await` if they return `Task<T>`. Otherwise they can be converted to `Task<T>` or left as-is

Reply:
In order to do that, I would need wrap in a ValueTask.

Reply:
There's a method like `AsTask()` or something like that if you wanted